### ISSUE TITLE
Corrupted db due to disk error - index out of range / failed to get all reachable pages

### ISSUE THREAD

**Issue by sangaman** [etcd-io/bbolt#197]
I have a ~200 MB bbolt db that was corrupted recently due to a failing SSD. When I noticed the disk was failing, I powered down the machine, removed the disk and used `ddrescue` to clone it to a new disk. `ddrescue` reported it was able to rescue 99.99% of the data from the failing disk. Most applications/data worked fine on the cloned disk, however the bbolt db appears to be corrupted.

Attempting to use the corrupted db with its application results in:

```
2019-12-20 12:37:31.963 [INF] LTND: Version: 0.8.2-beta commit=queue/v1.0.2-138-g6e0cfe579b3c884359b2a11a2130f2f41ebeec28-dirty, build=production, logging=default
2019-12-20 12:37:31.963 [INF] LTND: Active chain: Bitcoin (network=mainnet)
panic: freepages: failed to get all reachable pages (page 0: invalid type: unknown<00>)

goroutine 56 [running]:
github.com/coreos/bbolt.(*DB).freepages.func2(0x1a18640)
        /home/bitcoin/go/pkg/mod/github.com/coreos/bbolt@v1.3.3/db.go:1003 +0xcc
created by github.com/coreos/bbolt.(*DB).freepages
        /home/bitcoin/go/pkg/mod/github.com/coreos/bbolt@v1.3.3/db.go:1001 +0x108
```

Running `bolt check` or `bolt compact` on it fails with:

```
panic: runtime error: index out of range [18446744073709547520] with length 2147483647

goroutine 1 [running]:
github.com/boltdb/bolt.(*DB).page(...)
        /home/bitcoin/go/src/github.com/boltdb/bolt/db.go:794
github.com/boltdb/bolt.Open(0xbe83f0bb, 0xa, 0x1b6, 0x201238, 0x0, 0x0, 0x41a090)
        /home/bitcoin/go/src/github.com/boltdb/bolt/db.go:237 +0x544
main.(*CheckCommand).Run(0x494f28, 0x40a170, 0x1, 0x2, 0x0, 0x0)
        /home/bitcoin/go/src/github.com/boltdb/bolt/cmd/bolt/main.go:182 +0x150
main.(*Main).Run(0x43278c, 0x40a168, 0x2, 0x3, 0x0, 0x14c01)
        /home/bitcoin/go/src/github.com/boltdb/bolt/cmd/bolt/main.go:104 +0x530
main.main()
        /home/bitcoin/go/src/github.com/boltdb/bolt/cmd/bolt/main.go:64 +0xa
```

Might there be any way to repair the database or to recover data from it? I suspect that the majority of the data is intact, but that data loss/corruption in certain parts of the file is resulting in panics.

Related issue: https://github.com/lightningnetwork/lnd/issues/3867

**Comment by nonpcnpc** [comment]
does the new disk have the same page size as the old one?

**Comment by sangaman** [comment]
I'm not certain about page size, I believe so because I cloned the disk including the partition tables. I didn't copy the file from the old disk because it was failing and throwing filesystem errors - from my understanding trying to use a disk like that normally can lead to more corruption and I wasn't sure the filesystem could be repaired on the failing disk. Cloning the disk with ddrescue attempts to scrape data from bad sectors multiple times which I also thought would be handy.

I did wind up recovering the database, however:

> I used a modified version of bbolt where I commented out the line that was triggering the panic to rung bbolt compact on my corrupted db. The compacted db now passes bbolt check

That line was https://github.com/etcd-io/bbolt/blob/master/db.go#L1003

There was some data loss/corruption within the database and I needed to take a few more steps to get it to work with the application, but I retained most of the important data.
